File: zmqReqRepBrokerServer/nodejs-server/worker.js
// Connects REP socket to tcp://*:5560
var zmq = require('zmq'),
    responder = zmq.socket('rep');

responder.identity = 'node.js server [' + process.pid + ']'
responder.connect('tcp://localhost:5560');
responder.on('message', function(msg) {
    delay = getRandomInt(0, 4)
    console.log('Delay ', delay, ' seconds', ' -- received request:', msg.toString());
    setTimeout(function() {
        responder.send(responder.identity + ' echo: ' + msg);
    }, delay * 1000);
});

// Returns a random integer between min (included) and max (excluded)
// // Using Math.round() will give you a non-uniform distribution!
function getRandomInt(min, max) {
    return Math.floor(Math.random() * (max - min)) + min;
}

File: zmqReqRepBrokerServer/py-server/worker.py
import os
import zmq
import random
import time
# https://www.digitalocean.com/community/tutorials/how-to-work-with-the-zeromq-messaging-library

# ZeroMQ Context
context = zmq.Context()

# Define the socket using the "Context"
sock = context.socket(zmq.REP)
sock.connect("tcp://127.0.0.1:5560")

pid = os.getpid()

# Run a simple "Echo" server
while True:
    message = sock.recv()
    delay = random.randint(0, 3)
    print "Delay " + str(delay) + " seconds" + "Recv: " + message
    time.sleep(delay)
    sock.send("Python server [" + str(pid) + "] echo: " + message)
    #print "Echo: " + message
